File: piquasso/_math/permanent.py
#
# Copyright 2021-2024 Budapest Quantum Computing Group
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import numpy as np

import numba as nb

from numba import int64

from piquasso._math.combinatorics import comb


@nb.njit(cache=True, parallel=True)
def permanent(matrix, rows, cols):
    """Calculates the permanent of a matrix given row and column repetitions.

    Translated from PiquassoBoost, original implementation:
    https://github.com/Budapest-Quantum-Computing-Group/piquassoboost/blob/main/piquassoboost/sampling/source/BBFGPermanentCalculatorRepeated.hpp

    Implements Eq. (8) from https://arxiv.org/pdf/2309.07027.pdf.
    """  # noqa: E501

    rows = rows.astype(np.int64)
    cols = cols.astype(np.int64)

    # Determine minimal nonzero element
    min_idx = 0
    minelem = 0
    for i in range(len(rows)):
        if minelem == 0 or rows[i] < minelem and rows[i] != 0:
            minelem = rows[i]
            min_idx = i

    if len(rows) > 0 and minelem != 0:
        rows_ = np.empty(len(rows) + 1, dtype=np.int64)
        rows_[0] = 1
        rows_[1:] = rows
        rows_[1 + min_idx] -= 1

        matrix_ = np.empty(
            shape=(matrix.shape[0] + 1, matrix.shape[1]), dtype=matrix.dtype
        )

        matrix_[0] = matrix[min_idx]

        matrix_[1:] = matrix
        rows = rows_
        matrix = matrix_

    sum_rows = np.sum(rows)
    sum_cols = np.sum(cols)

    if matrix.shape[0] == 0 or matrix.shape[1] == 0 or sum_rows == 0 or sum_cols == 0:
        return 1.0

    if matrix.shape[0] == 1:
        ret = 1.0
        for idx in range(len(cols)):
            for _ in range(cols[idx]):
                ret *= matrix[idx, 0]

        return ret

    mtx2 = matrix * 2

    n_ary_limits = np.empty(len(rows) - 1, dtype=np.int64)

    for idx in range(len(n_ary_limits)):
        n_ary_limits[idx] = rows[idx + 1] + 1

    idx_max = n_ary_limits[0]
    for idx in range(1, len(n_ary_limits)):
        idx_max *= n_ary_limits[idx]

    nthreads = nb.config.NUMBA_NUM_THREADS

    concurrency = min(nthreads, idx_max, 32)

    permanent = 0.0

    for job_idx in nb.prange(concurrency):
        partial_permanent = 0.0

        work_batch = idx_max // concurrency
        initial_offset = job_idx * work_batch
        offset_max = (job_idx + 1) * work_batch - 1
        if job_idx == concurrency - 1:
            offset_max = idx_max - 1

        gcode_counter = NaryGrayCodeCounter(n_ary_limits, initial_offset)

        gcode_counter.offset_max = offset_max
        gcode = gcode_counter.gray_code
        binomial_coeff = 1

        colsum = np.copy(matrix[0])

        minus_signs_all = 0

        row_idx = 1

        for idx in range(len(gcode)):
            minus_signs = gcode[idx]
            rows_current = rows[idx + 1]

            for col_idx in range(len(cols)):
                colsum[col_idx] += matrix[row_idx, col_idx] * (
                    rows_current - 2 * minus_signs
                )

            minus_signs_all += minus_signs

            binomial_coeff *= comb(rows_current, minus_signs)

            row_idx += 1

        parity = 1 if (minus_signs_all % 2 == 0) else -1

        colsum_prod = parity
        for idx in range(len(cols)):
            for _ in range(cols[idx]):
                colsum_prod *= colsum[idx]

        partial_permanent += colsum_prod * binomial_coeff

        for idx in range(initial_offset + 1, offset_max + 1):
            flag, changed_index, value_prev, value = gcode_counter.next()
            if flag:
                break

            parity = -parity

            row_offset = changed_index + 1
            colsum_prod = parity
            for col_idx in range(len(cols)):
                if value_prev < value:
                    colsum[col_idx] -= mtx2[row_offset, col_idx]

                else:
                    colsum[col_idx] += mtx2[row_offset, col_idx]

                for _ in range(cols[col_idx]):
                    colsum_prod *= colsum[col_idx]

            rows_current = rows[changed_index + 1]
            binomial_coeff = (
                (binomial_coeff * value_prev / (rows_current - value))
                if value < value_prev
                else (binomial_coeff * (rows_current - value_prev) / value)
            )

            partial_permanent += colsum_prod * binomial_coeff

        permanent += partial_permanent

    permanent /= 2 ** (sum_rows - 1)

    return permanent


@nb.experimental.jitclass(
    [
        ("gray_code", int64[:]),
        ("n_ary_limits", int64[:]),
        ("counter_chain", int64[:]),
        ("offset_max", int64),
        ("offset", int64),
    ]
)
class NaryGrayCodeCounter(object):
    def __init__(self, n_ary_limits_in, initial_offset):
        self.n_ary_limits = np.copy(n_ary_limits_in)
        if len(self.n_ary_limits) == 0:
            self.offset_max = 0
            self.offset = 0
            return

        self.offset_max = self.n_ary_limits[0]
        for idx in range(1, len(self.n_ary_limits)):
            self.offset_max *= self.n_ary_limits[idx]

        self.offset_max -= 1
        self.offset = initial_offset

        self._initialize(initial_offset)

    def _initialize(self, initial_offset):
        if initial_offset < 0 or initial_offset > self.offset_max:
            raise

        self.counter_chain = np.empty_like(self.n_ary_limits)

        for idx in range(len(self.n_ary_limits)):
            self.counter_chain[idx] = initial_offset % self.n_ary_limits[idx]
            initial_offset /= self.n_ary_limits[idx]

        self.gray_code = np.empty_like(self.n_ary_limits)
        parity = 0
        for jdx in range(len(self.n_ary_limits) - 1, -1, -1):
            self.gray_code[jdx] = (
                self.n_ary_limits[jdx] - 1 - self.counter_chain[jdx]
                if parity
                else self.counter_chain[jdx]
            )
            parity = parity ^ (self.gray_code[jdx] & 1)

    def next(self):
        changed_index = 0

        if self.offset >= self.offset_max:
            return True, 0, 0, 0

        update_counter = True
        counter_chain_idx = 0
        while update_counter:

            if (
                self.counter_chain[counter_chain_idx]
                < self.n_ary_limits[counter_chain_idx] - 1
            ):
                self.counter_chain[counter_chain_idx] += 1
                update_counter = False

            elif (
                self.counter_chain[counter_chain_idx]
                == self.n_ary_limits[counter_chain_idx] - 1
            ):
                self.counter_chain[counter_chain_idx] = 0
                update_counter = True

            counter_chain_idx += 1

        parity = 0
        for jdx in range(len(self.n_ary_limits) - 1, -1, -1):
            gray_code_new_val = (
                self.n_ary_limits[jdx] - 1 - self.counter_chain[jdx]
                if parity
                else self.counter_chain[jdx]
            )
            parity = parity ^ (gray_code_new_val & 1)

            if gray_code_new_val != self.gray_code[jdx]:
                value_prev = self.gray_code[jdx]
                value = gray_code_new_val
                self.gray_code[jdx] = gray_code_new_val
                changed_index = jdx
                break

        self.offset += 1

        return False, changed_index, value_prev, value

File: src/permanent.cpp
/*
 * Copyright 2021-2025 Budapest Quantum Computing Group
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "matrix.hpp"
#include "utils.hpp"
#include <cmath>
#include <complex>
#include <cstdint>
#include <vector>

#include "n_aryGrayCodeCounter.hpp"
#include <thread>
#if defined(_OPENMP)
#include <omp.h>
#endif

/**
 * @brief Computes the permanent of a given matrix.
 *
 * This function calculates the permanent of the input matrix given row and column
 * repetitions.
 *
 * This takes inspiration from the original PiquassoBoost implementation found in
 * https://github.com/Budapest-Quantum-Computing-Group/piquassoboost/blob/main/piquassoboost/sampling/source/BBFGPermanentCalculatorRepeated.hpp
 * which is also reimplemented in
 * https://github.com/0xSooki/permanent-boost.
 *
 * The formula used is from Eq. (8) in https://arxiv.org/pdf/2309.07027.pdf.
 *
 * @tparam TScalar The type of the elements in the matrix.
 * @param A The input matrix for which the permanent is to be computed.
 * @param rows A vector representing the multiplicity of each row.
 * @param cols A vector representing the multiplicity of each column.
 * @return The permanent of the input matrix.
 */
template <typename T>
std::complex<T> permanent_cpp(Matrix<std::complex<T>> &A, Vector<int> &rows, Vector<int> &cols)
{
    using TComplex = std::complex<T>;

    size_t min_idx = 0;
    int minelem = 0;

    for (size_t i = 0; i < rows.size(); i++)
    {
        if (minelem == 0 || (rows[i] < minelem && rows[i] != 0))
        {
            minelem = rows[i];
            min_idx = i;
        }
    }

    if (rows.size() > 0 && minelem != 0)
    {
        Vector<int> rows_(rows.size() + 1);

        rows_[0] = 1;

        for (size_t i = 0; i < rows.size(); i++)
        {
            rows_[i + 1] = rows[i];
        }
        rows_[1 + min_idx] -= 1;

        Matrix<TComplex> mtx_(A.rows + 1, A.cols);

        for (size_t j = 0; j < A.cols; j++)
        {
            mtx_(0, j) = A(min_idx, j);
        }

        for (size_t i = 0; i < A.rows; i++)
        {
            for (size_t j = 0; j < A.cols; j++)
            {
                mtx_(i + 1, j) = A(i, j);
            }
        }

        rows = rows_;
        A = mtx_;
    }

    int sum_rows = rows.sum();
    int sum_cols = cols.sum();

    if (sum_rows != sum_cols)
    {
        std::string error("Number of input and output states should be equal");
        throw error;
    }

    if (A.rows == 0 || A.cols == 0 || sum_rows == 0 || sum_cols == 0)
        // the permanent of an empty matrix is 1 by definition
        return TComplex(1.0, 0.0);

    if (A.rows == 1)
    {
        TComplex ret(1.0, 0.0);
        for (size_t i = 0; i < cols.size(); i++)
        {
            for (int j = 0; j < cols[i]; j++)
            {
                ret *= A[i];
            }
        }

        return TComplex(ret.real(), ret.imag());
    }

    Matrix<TComplex> mtx2(A.rows, A.cols);
    for (size_t i = 0; i < A.size(); i++)
    {
        mtx2[i] = A[i] * TComplex(2.0, 0.0);
    }

    // Create a dynamically allocated array for n-ary limits.
    size_t n_ary_size = rows.size() - 1;
    int *n_ary_limits = new int[n_ary_size];
    for (size_t i = 0; i < n_ary_size; i++)
    {
        n_ary_limits[i] = rows[i + 1] + 1;
    }

    int64_t idx_max = n_ary_limits[0];
    for (size_t i = 1; i < n_ary_size; i++)
    {
        idx_max *= n_ary_limits[i];
    }

    // determine the concurrency of the calculation
    unsigned int n_threads = std::thread::hardware_concurrency();
    auto concurrency = static_cast<int64_t>(n_threads * 4);
    concurrency = concurrency < idx_max ? concurrency : idx_max;

    std::vector<TComplex> thread_results(static_cast<unsigned int>(concurrency), TComplex(0.0, 0.0));

// See: https://stackoverflow.com/a/8447025
#if defined(_OPENMP)
#pragma omp parallel for num_threads(static_cast<unsigned int>(concurrency))
#endif
    for (int64_t job_idx = 0; job_idx < concurrency; job_idx++)
    {
        // initial offset and upper boundary of the gray code counter
        int64_t work_batch = idx_max / concurrency;
        int64_t initial_offset = job_idx * work_batch;
        int64_t offset_max = (job_idx + 1) * work_batch - 1;
        if (job_idx == concurrency - 1)
        {
            offset_max = idx_max - 1;
        }

        // Use the updated gray code counter, passing both the limits pointer and its size.
        n_aryGrayCodeCounter gcode_counter(n_ary_limits, n_ary_size, initial_offset);
        gcode_counter.set_offset_max(offset_max);

        // the gray code vector
        int *gcode = gcode_counter.get();

        // calculate the initial column sum and binomial coefficient
        int binomial_coeff = 1;

        Matrix<TComplex> colsum(1, cols.size());
        std::uninitialized_copy_n(A.data, colsum.size(), colsum.data);

        // variable to count all the -1 elements in the delta vector
        int minus_signs_all = 0;

        for (size_t i = 0; i < n_ary_size; i++)
        {
            // the value of the element of the gray code stands for the number of
            // \delta_i=-1 elements in the subset of multiplicated rows
            const int &minus_signs = gcode[i];
            int row_mult_current = rows[i + 1];

            for (size_t j = 0; j < cols.size(); j++)
            {
                colsum[j] += static_cast<TComplex>(A(i + 1, j)) *
                             static_cast<T>(row_mult_current - 2 * minus_signs);
            }

            minus_signs_all += minus_signs;

            // update the binomial coefficient
            binomial_coeff *= binomialCoeff<int>(row_mult_current, minus_signs);
        }

        // variable to refer to the parity of the delta vector (+1 if even, -1 if odd)
        char parity = (minus_signs_all % 2 == 0) ? 1 : -1;

        TComplex colsum_prod(static_cast<T>(parity), static_cast<T>(0.0));
        for (size_t i = 0; i < cols.size(); i++)
        {
            for (int j = 0; j < cols[i]; j++)
            {
                colsum_prod *= colsum[i];
            }
        }

        // add the initial addend to the permanent - store in thread-local result
        TComplex &addend_loc = thread_results[static_cast<unsigned int>(job_idx)];
        addend_loc += colsum_prod * static_cast<T>(binomial_coeff);

        // iterate over gray codes to calculate permanent addends
        for (int64_t i = initial_offset + 1; i < offset_max + 1; i++)
        {
            int changed_index = 0;
            int prev_value = 0;
            int value = 0;
            if (gcode_counter.next(changed_index, prev_value, value))
            {
                break;
            }

            parity = -parity;

            // update column sum and calculate the product of the elements
            int row_offset = changed_index + 1;
            colsum_prod = TComplex(static_cast<T>(parity), static_cast<T>(0.0));
            for (size_t j = 0; j < cols.size(); j++)
            {
                colsum[j] += mtx2(row_offset, j) *
                             static_cast<T>(prev_value - value);
                for (int k = 0; k < cols[j]; k++)
                {
                    colsum_prod *= colsum[j];
                }
            }

            int row_mult_current = rows[changed_index + 1];
            binomial_coeff =
                value < prev_value
                    ? binomial_coeff * prev_value / (row_mult_current - value)
                    : binomial_coeff * (row_mult_current - prev_value) / value;

            addend_loc += colsum_prod * static_cast<T>(binomial_coeff);
        }
    }

    TComplex permanent(0.0, 0.0);
    for (const auto &result : thread_results)
    {
        permanent += result;
    }

    permanent /= static_cast<T>(ldexp(1.0, sum_rows - 1));

    delete[] n_ary_limits;

    return permanent;
}

template std::complex<float> permanent_cpp<float>(
    Matrix<std::complex<float>> &matrix, Vector<int> &rows, Vector<int> &cols);
template std::complex<double> permanent_cpp<double>(
    Matrix<std::complex<double>> &matrix, Vector<int> &rows, Vector<int> &cols);
